corehq/apps/userreports/reports/filters/choice_providers.py

```diff
@@ -211,6 +213,20 @@ def get_choices_for_known_values(self, values, user):
     def default_value(self, user):
         return None
 
+    def _apply_query_filter(self, query, query_context):
+        query = query.filter(self._sql_column.ilike("%{}%".format(query_context.query)))
+        # We need to sort this by relevance, otherwise we might
+        # be pushing the most exact matches to the end of the list
+        # and cut it off by applying the limit
+        return query.distinct().order_by(
+            case([
+                (self._sql_column.ilike(query_context.query), 0),  # Exact matches first
+                (self._sql_column.ilike(f"{query_context.query}%"), 1),  # Prefix matches next
```

Comment:
I have not used case before so I am not fully familiar with the change here.

Just one thing to add, I assume this adds considerable load, so we could just limit to exact match and substring.
Or, is it possible to fetch all matches and then apply some filter, sort & limit in python instead? I wonder if we are making the db do a lot of work in the query and so instead could just do that in python since the db isn't getting any advantage of the limit anyway.

Reply:
> we could just limit to exact match and substring

Sounds reasonable.

Thanks for the input. I'm slightly hesitant to push the operations into python itself, since DBs are very efficient and optimised for handling calculations over large datasets. Let's see what others say as well.